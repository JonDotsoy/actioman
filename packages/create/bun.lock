{
  "lockfileVersion": 1,
  "workspaces": {
    "": {
      "name": "create-actioman",
      "dependencies": {
        "@jondotsoy/flags": "^2.3.0",
      },
      "devDependencies": {
        "@types/bun": "latest",
      },
      "peerDependencies": {
        "typescript": "^5.0.0",
      },
    },
  },
  "packages": {
    "@jondotsoy/console-draw": ["@jondotsoy/console-draw@0.1.2", "", { "dependencies": { "@jondotsoy/style-text": "^0.9.1" }, "peerDependencies": { "prettier": "^3.3.3", "typescript": "^5.0.0" } }, "sha512-BBZGLd6RzjKxKcgD2OwD7v6GeKlf53Ysk59H+7+HvyGLcJZ7AJk8rOCq/yv813KbWV2teTCEBQXnnP5hZe/2CA=="],

    "@jondotsoy/flags": ["@jondotsoy/flags@2.3.0", "", { "dependencies": { "@jondotsoy/console-draw": "^0.1.2" } }, "sha512-vUMe4NdXKv8RxMXrKCJLsHBt3XQ8/YV4Fyr1I43vSLe5v0M8sdY3wJl1qdTgwRBZES2zMLoenxOUSNGE9r/5NA=="],

    "@jondotsoy/style-text": ["@jondotsoy/style-text@0.9.2", "", { "peerDependencies": { "typescript": "^5.0.0" } }, "sha512-6fonOzrANwrw+UIM4fivtDZLvuCN00yH0lCEYi14ln4j8fSHZokb6TLrqOu38GCqqQnZ2LuzJTQBWgv7jN6iVA=="],

    "@types/bun": ["@types/bun@1.2.2", "", { "dependencies": { "bun-types": "1.2.2" } }, "sha512-tr74gdku+AEDN5ergNiBnplr7hpDp3V1h7fqI2GcR/rsUaM39jpSeKH0TFibRvU0KwniRx5POgaYnaXbk0hU+w=="],

    "@types/node": ["@types/node@22.13.1", "", { "dependencies": { "undici-types": "~6.20.0" } }, "sha512-jK8uzQlrvXqEU91UxiK5J7pKHyzgnI1Qnl0QDHIgVGuolJhRb9EEl28Cj9b3rGR8B2lhFCtvIm5os8lFnO/1Ew=="],

    "@types/ws": ["@types/ws@8.5.14", "", { "dependencies": { "@types/node": "*" } }, "sha512-bd/YFLW+URhBzMXurx7lWByOu+xzU9+kb3RboOteXYDfW+tr+JZa99OyNmPINEGB/ahzKrEuc8rcv4gnpJmxTw=="],

    "bun-types": ["bun-types@1.2.2", "", { "dependencies": { "@types/node": "*", "@types/ws": "~8.5.10" } }, "sha512-RCbMH5elr9gjgDGDhkTTugA21XtJAy/9jkKe/G3WR2q17VPGhcquf9Sir6uay9iW+7P/BV0CAHA1XlHXMAVKHg=="],

    "prettier": ["prettier@3.4.2", "", { "bin": { "prettier": "bin/prettier.cjs" } }, "sha512-e9MewbtFo+Fevyuxn/4rrcDAaq0IYxPGLvObpQjiZBMAzB9IGmzlnG9RZy3FFas+eBMu2vA0CszMeduow5dIuQ=="],

    "typescript": ["typescript@5.7.3", "", { "bin": { "tsc": "bin/tsc", "tsserver": "bin/tsserver" } }, "sha512-84MVSjMEHP+FQRPy3pX9sTVV/INIex71s9TL2Gm5FG/WG1SqXeKyZ0k7/blY/4FdOzI12CBy1vGc4og/eus0fw=="],

    "undici-types": ["undici-types@6.20.0", "", {}, "sha512-Ny6QZ2Nju20vw1SRHe3d9jVu6gJ+4e3+MMpqu7pqE5HT6WsTSlce++GQmK5UXS8mzV8DSYHrQH+Xrf2jVcuKNg=="],
  }
}
