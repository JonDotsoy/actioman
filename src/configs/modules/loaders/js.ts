export default async (sourceLocation: string) => {
  const { default: def } = await import(sourceLocation);
  return def;
};
