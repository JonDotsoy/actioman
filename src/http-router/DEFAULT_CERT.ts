export const DEFAULT_CERT =
  "-----BEGIN CERTIFICATE-----\n" +
  "MIIDCTCCAfGgAwIBAgIUcx4/uMJz+em1t9pyg8YBn8mOtz8wDQYJKoZIhvcNAQEL\n" +
  "BQAwFDESMBAGA1UEAwwJbG9jYWxob3N0MB4XDTI1MDIwNzE1MDcyMloXDTI1MDMw\n" +
  "OTE1MDcyMlowFDESMBAGA1UEAwwJbG9jYWxob3N0MIIBIjANBgkqhkiG9w0BAQEF\n" +
  "AAOCAQ8AMIIBCgKCAQEApcFu5mawUooMhsH/1bYdRMn1PTlmRiPRJkoookkWOhVH\n" +
  "703iYVCkZL39DF9tXG4iANNdjFepVIcSZduyZBXslAvuvFWaNu9I52IYyDYDadDT\n" +
  "oLvXFi3dsfnIrGf68JLBqZnOSxVLg80Ht3maOSbc36DbUO5YE9nU0nhA/wQ983j0\n" +
  "FDzBKzrPYPFNIbmI3SEVPjhrFrxci7+Tbdky4SpIwLRN7SWpbu/GPmLzlvD8XMuW\n" +
  "cw1dsBoqV9w+pMwy2DTctQ3SU3nbxuvMgMKu4wuX9mWLiCmykGcLvYUrI8CKFzzP\n" +
  "f48Qd2/HC1lzwCKiwn57MhokbzcKWhB8E96X3HGtVwIDAQABo1MwUTAdBgNVHQ4E\n" +
  "FgQUPsTW/85jbxCCvyiMwx33xM6nh/IwHwYDVR0jBBgwFoAUPsTW/85jbxCCvyiM\n" +
  "wx33xM6nh/IwDwYDVR0TAQH/BAUwAwEB/zANBgkqhkiG9w0BAQsFAAOCAQEANfYP\n" +
  "er5WrTFuWtY2ApxMjvDWo0xuL6+kCv4UkaLtfsR4TkcF9svGtb/PvrlOmffP+By1\n" +
  "TkBoMju3uHU7/hPo5UMRUI445QnIVSSpLaHnLHBi85KGMl55ErRbOGVgYR1n+Qxo\n" +
  "nFwucn2hjHBmswXNY9DJx8vmewKqa9Kc694YxNx44tmMNbokUYKliofnisxxhM3Y\n" +
  "4REA+w3bme7z7JDLt0J1EGCAEeT2NcLbnXw8gEKJ0D39Zt5b5qR1Re4wWtknXlkc\n" +
  "5VGw96c5jbejD49BJsDWqVsJG1D4dhJ61ybBT6j2aRnSH+MxPHpH/Rm9wXIGha4G\n" +
  "mZLuodbHDBrL/FT4cQ==\n" +
  "-----END CERTIFICATE-----\n";
