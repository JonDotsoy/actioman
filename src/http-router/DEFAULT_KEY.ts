if (process.env.NODE_ENV !== "test") {
  console.warn("The variable DEFAULT_KEY is only to test environment");
}

export const DEFAULT_KEY =
  "-----BEGIN PRIVATE KEY-----\n" +
  "MIIEvAIBADANBgkqhkiG9w0BAQEFAASCBKYwggSiAgEAAoIBAQClwW7mZrBSigyG\n" +
  "wf/Vth1EyfU9OWZGI9EmSiiiSRY6FUfvTeJhUKRkvf0MX21cbiIA012MV6lUhxJl\n" +
  "27JkFeyUC+68VZo270jnYhjINgNp0NOgu9cWLd2x+cisZ/rwksGpmc5LFUuDzQe3\n" +
  "eZo5JtzfoNtQ7lgT2dTSeED/BD3zePQUPMErOs9g8U0huYjdIRU+OGsWvFyLv5Nt\n" +
  "2TLhKkjAtE3tJalu78Y+YvOW8Pxcy5ZzDV2wGipX3D6kzDLYNNy1DdJTedvG68yA\n" +
  "wq7jC5f2ZYuIKbKQZwu9hSsjwIoXPM9/jxB3b8cLWXPAIqLCfnsyGiRvNwpaEHwT\n" +
  "3pfcca1XAgMBAAECggEAGE8voAvEZT6IysSt7B5a2YlS9aky5it/k4GBxxc38snH\n" +
  "NIH0JviIO5wffg6Kv+kJBJz7FcePdm2e1aBsX8BVY3sChs15IgBi7UhWQRqm0oCF\n" +
  "YQfbag3M+zqaYyqWw2yKiYCcoCdd3VUXAiIL68DW2NdDAvIJqdOjIkDvq1bobQn+\n" +
  "MNdNBnu58t95JRfEbQHiDUUVTGv1b4E8gAUtCmsDPtsSQAox8rTD4UTA0+9oxE5e\n" +
  "SoV69YWS+BqFaDZgxxTKxyoL6g4K7WrF58oOSIlayUBGOO0VpGBf3sFB9YvYw76P\n" +
  "7m8S1EY71G8/Bp4JoVrNR7TemG0Qotc/3AapEjyv7QKBgQDVKfOWe81oKuFJlpKP\n" +
  "1W0v7xoo0fV9GXJlOrpDTvcuoY+XiMuQtTWDl7de+4y+mzQ+YUWKCrEn6yj0eUgy\n" +
  "owEpk8p6E+nvU0w9Yhnlk0wmfHkQw/P4kwCuG7sFvcpJ8471Ad94NWLddeHdVQZb\n" +
  "3A8PZgXouoBstTlKJ9WkZHxmCwKBgQDHEJmiR/U6m882A8G7yhoDB4U2tYPS2WZq\n" +
  "sZdgaCDrYZNME46N11b6HEJYD6jHyP33FPgyLD9uz157/1WTod6y2fdrlRnTWXZt\n" +
  "KKX837tOaKZFYpJtTzVFdR6nCO6O4fL6tK/CyNIWOB+kAXmtr8GBfbBfGHuePuYr\n" +
  "9jjkGmIhZQKBgBes7Cgnof3S3Wjd93NQQitTxDUKdNirLeB+cFevepbMTPWItQti\n" +
  "hQoTnMNI7y3wcRIInb+iRmDnponYuqMdJbDngFKkyZPSTaJ3teagwJF5u7yX9w4B\n" +
  "zc6It/IATWp/Sb5+scJZM5zjTe3WVrl3LJkKPNfHdyIb0rwgxoLWwlZXAoGALI8s\n" +
  "tMTCwgHKml6w6SZOOEtgqGRtQ69iW5OxPOVSKR5IaIkgKeQAFjBpMWkf8MivALXy\n" +
  "8Ngyws0sazk2oLtlLEfaOHzBsSufdMPalAZI2T5XqcRs3ocSd5/F4sH/FIDDcRct\n" +
  "GL73N/4MJxwlbpmfCStnZYIHed6+iyysg3LccHkCgYBTPy0YejvOaTr3FMqIeL4r\n" +
  "3TsSbN9Q3/BX0oEvgTiJvLU1wsT3G0VzBE9qDbAmn9NuJhDKKgQRA7BdPVijZwib\n" +
  "NS49DtF2a5rB2wXfWNTG4KsQUPQ+zkhu3TyFUQijdlD+Vh47vD3bnH8hgInllUNc\n" +
  "WWPRkIP/TR8122xeYnv/hQ==\n" +
  "-----END PRIVATE KEY-----\n";
