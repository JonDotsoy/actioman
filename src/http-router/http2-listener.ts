import { HTTPRouter } from "./http-router.js";
import type { ConfigsModule } from "../configs/configs.js";
import { sanitizeHostname } from "./utils/sanitize-hostname.js";
import { findNextPort } from "./utils/find-next-port.js";
import * as http2 from "http2";
import { DEFAULT_CERT } from "./DEFAULT_CERT.js";
import { DEFAULT_KEY } from "./DEFAULT_KEY.js";
import { requestHttp2ToRequest } from "./utils/request-http2-to-request.js";

type ListenOptions = {
  silent: boolean;
};

export const INITIAL_PORT = 30_320;

export class HTTP2Lister {
  server: http2.Http2Server;
  url: URL | undefined;

  constructor(readonly httpRouter: HTTPRouter) {
    this.server = http2.createSecureServer(
      {
        key: DEFAULT_KEY,
        cert: DEFAULT_CERT,
        settings: { initialWindowSize: 8 * 1024 * 1024 },
      },
      async (requestHttp2, responseHttp2) => {
        try {
          // console.log(r)

          const response = await this.httpRouter.router.fetch(
            requestHttp2ToRequest(requestHttp2),
          );

          if (!response) {
            responseHttp2.statusCode = 404;
            return responseHttp2.end();
          }

          responseHttp2.statusCode = response.status;
          response.headers.forEach((value, key) => {
            if (key.startsWith(":")) return;
            if (responseHttp2.hasHeader(key)) {
              responseHttp2.appendHeader(key, value);
            } else {
              responseHttp2.setHeader(key, value);
            }
          });
          if (response.body) {
            const reader = response.body.getReader();
            for (
              let readResult = await reader.read();
              !readResult.done && !requestHttp2.aborted;
              readResult = await reader.read()
            ) {
              responseHttp2.write(readResult.value);
            }
          }
          return responseHttp2.end();
        } catch (error) {
          console.error(error);
          responseHttp2.statusCode = 500;
          return responseHttp2.end();
        }
      },
    );
  }

  async close() {
    await new Promise<null>((resolve, reject) => {
      if (this.server.address() === null) return resolve(null);
      this.server.close();
      return resolve(null);
    });
  }

  async listen(port?: number, hostname?: string, options?: ListenOptions) {
    const silent = options?.silent ?? HTTP2Lister.defaultOptions.silent;

    const log = (message: string) => {
      if (silent) return;
      console.log(message);
    };

    const portToListen = port ?? (await findNextPort());
    const hostnameToListen = hostname ?? "::";

    const url = await new Promise<URL>((resolve) => {
      this.server.listen(portToListen, hostnameToListen, () => {
        const address = this.server.address();
        if (typeof address === "object" && address !== null)
          return resolve(
            new URL(
              `https://${sanitizeHostname(address.address)}:${address.port}`,
            ),
          );
        return resolve(
          new URL(
            `https://${sanitizeHostname(hostnameToListen)}:${portToListen}`,
          ),
        );
      });
    });

    if (!silent) {
      for (const route of this.httpRouter.router.routes) {
        const method = route.method;
        const urlPattern = route.urlPattern;
        log(`Route ${method} ${urlPattern.pathname}`);
      }
    }
    log(`Listening on ${url.toString()}`);

    this.url = url;

    return url;
  }

  static fromModule(module: unknown, configs?: ConfigsModule) {
    const httpRouter = HTTPRouter.fromModule(module, configs);
    if (!httpRouter) throw new Error("No actions found");
    return new HTTP2Lister(httpRouter);
  }

  static defaultOptions = {
    silent: false,
  };
}

export const HTTPLister = HTTP2Lister;
