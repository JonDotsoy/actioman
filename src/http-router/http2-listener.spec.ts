import { CleanupTasks } from "@jondotsoy/utils-js/cleanuptasks";
import { describe, expect, it } from "bun:test";
import { HTTP2Lister } from "./http2-listener.js";
import * as http2 from "http2";
import { DEFAULT_CERT } from "./DEFAULT_CERT.js";
import { DEFAULT_KEY } from "./DEFAULT_KEY.js";

describe("async HTTP2Lister", () => {
  it("should return 200 for /__actions", async () => {
    await using cleanupTasks = new CleanupTasks();

    const http2Lister = HTTP2Lister.fromModule(
      {
        hi: () => "ok",
      },
      {
        server: {
          ssl: {
            key: DEFAULT_KEY,
            cert: DEFAULT_CERT,
          },
        },
      },
    );
    cleanupTasks.add(() => http2Lister.close());

    const url = await http2Lister.listen();

    const client = http2.connect(url, { ca: DEFAULT_CERT });
    cleanupTasks.add(() => client.close());

    const { status, body } = await new Promise<{ body: any; status: number }>(
      (resolve, reject) => {
        const data: number[] = [];
        let status: any;

        client
          .request({
            [http2.constants.HTTP2_HEADER_PATH]: "/__actions",
            [http2.constants.HTTP2_HEADER_METHOD]: "GET",
          })
          .addListener("error", (err) => reject(err))
          .addListener("response", (headers) => {
            status = headers[http2.constants.HTTP2_HEADER_STATUS];
          })
          .addListener("data", (d) => data.push(...d))
          .addListener("close", () => {
            resolve({
              status,
              body: new TextDecoder().decode(new Uint8Array(data)),
            });
          });
      },
    );

    expect(status).toEqual(200);
    expect(body).toMatchSnapshot();
  });

  it("should return 200 for /__actions without ssl", async () => {
    await using cleanupTasks = new CleanupTasks();

    const http2Lister = HTTP2Lister.fromModule({
      hi: () => "ok",
    });
    cleanupTasks.add(() => http2Lister.close());

    const url = await http2Lister.listen();

    const client = http2.connect(url);
    cleanupTasks.add(() => client.close());

    const { status, body } = await new Promise<{ body: any; status: number }>(
      (resolve, reject) => {
        const data: number[] = [];
        let status: any;

        client
          .request({
            [http2.constants.HTTP2_HEADER_PATH]: "/__actions",
            [http2.constants.HTTP2_HEADER_METHOD]: "GET",
          })
          .addListener("error", (err) => reject(err))
          .addListener("response", (headers) => {
            status = headers[http2.constants.HTTP2_HEADER_STATUS];
          })
          .addListener("data", (d) => data.push(...d))
          .addListener("close", () => {
            resolve({
              status,
              body: new TextDecoder().decode(new Uint8Array(data)),
            });
          });
      },
    );

    expect(status).toEqual(200);
    expect(body).toMatchSnapshot();
  });
});
