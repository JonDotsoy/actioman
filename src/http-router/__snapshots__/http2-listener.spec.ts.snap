// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`a a 1`] = `"{"actions":{"hi":{"description":null,"input":null,"output":null}}}"`;

exports[`async HTTP2Lister should return 200 for /__actions 1`] = `"{"actions":{"hi":{"description":null,"input":null,"output":null}}}"`;

exports[`async HTTP2Lister  1`] = `"{"actions":{"hi":{"description":null,"input":null,"output":null}}}"`;

exports[`async HTTP2Lister should return 200 for /__actions without ssl 1`] = `"{"actions":{"hi":{"description":null,"input":null,"output":null}}}"`;
